students/kristianjf/lesson09/mailroom_5.py

```diff
@@ -0,0 +1,221 @@
+#!/usr/bin/env python3
+import sys
+'''
+Goal:
+You work in the mail room at a local charity. Part of your job is to write \
+incredibly boring, repetitive emails thanking your donors for their generous \
+gifts. You are tired of doing this over and over again, so you’ve decided to \
+let Python help you out of a jam and do your work for you.
+'''
+'''
+It should have a data structure that holds a list of your donors and a history of \
+the amounts they have donated. This structure should be populated at first with at \
+least five donors, with between 1 and 3 donations each.
+'''
+# self.DATA_STRUCTURE = [('Jeff Bezos', [200, 500, 1000]), \
+#                     ('Bill Gates', [1000]), \
+#                     ('Steve Jobs', [20, 50, 100])]
+# self.DATA_STRUCTURE_DICTS = [{'name': name, 'donation': donation} for name, donation in self.DATA_STRUCTURE]
+# # self.DATA_STRUCTURE_DICT = [ \
+# #                         {'name': 'Jeff Bezos', 'donation': [200, 500, 1000]},\
+# #                         {'name': 'Bill Gates', 'donation': [1000]},\
+# #                         {'name': 'Steve Jobs', 'donation': [20, 50, 100]}\
+# #                         ]
+
+class Donor:
+    def __init__(self, name, donations):
+        self._name = name
+        self._donations = [] if donations is None else donations
+    @property
+    def name(self):
+        return self._name
+    @property
+    def donations(self):
+        return self._donations
+    def donate(self, new_donation):
+        self.donations.append(new_donation)
+    def metrics(self):
+        self.num_d = len(self.donations)
+        self.total_d = sum(self.donations)
+        self.avg_d = self.total_d / self.num_d
+
+class DonorsHandle:
+    DATA_STRUCTURE = [] #[('',[])]
+    DATA_STRUCTURE_DICTS = [] #[{'name': None, 'donation': None}]
+    def __init__(self):
+        pass
+    def add_donor(self, donor_object):
+        self.DATA_STRUCTURE.append((donor_object.name, donor_object.donations))
+        self.DATA_STRUCTURE_DICTS.append({'name': donor_object.name, \
```

Comment:
Why both? duplicate data!

But a bigger issue -- you have a Donor bject type -- you don't want to unpack it into a different data structure here, rather simply store the donor objects:

`self.donors.append(donor_object)`

or maybe make a nice key from the donor object ans store in a dic for fast lookup:

`self.donors[donor_object.name.lower().replace(' ','')] = donor_object`

or better yet:

`self.donors[donor_object.normalized_name] = donor_object`

where `normalized_name` is a property that computes a "normalized" name from the donor name.

This is all so that you can look up a name even if the capitalization, or spaces, or .. is different.